spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteDataFilesAction.java

```diff
@@ -195,6 +124,33 @@ public void testBinPackWithFilter() {
     assertEquals("Rows must match", expectedRecords, actualRecords);
   }
 
+  // A version of this using more complicated partitioning is in the Spark Extensions Module
+  @Test
+  public void testBinPackAfterPartitionChange() {
+    Table table = createTable();
+    writeRecords(20, SCALE, 4);
+    shouldHaveFiles(table, 20);
+    table.updateSpec().addField(Expressions.ref("c1")).commit();
+
+    List<Object[]> originalData = currentData();
+
+    RewriteDataFiles.Result result =
+        basicRewrite(table)
+            .option(SortStrategy.MIN_INPUT_FILES, "1")
+            .option(SortStrategy.MIN_FILE_SIZE_BYTES, Integer.toString(averageFileSize(table) + 1000))
+            .execute();
+
+    Assert.assertEquals("Should have 1 fileGroup because all files were not correctly partitioned",
+        result.rewriteResults().size(), 1);
+
+    List<Object[]> postRewriteData = currentData();
+    assertEquals("We shouldn't have changed the data", originalData, postRewriteData);
+
+    shouldHaveSnapshots(table, 2);
+    shouldHaveACleanCache(table);
+    shouldHaveFiles(table, 4);
+  }
```

Comment:
Question / thought: Would there be benefit in adding a test that starts out with an identity transform on one column, then switches to an identity transform on another column, and then binpack?

We wouldn't have the complication of having to refactor so much, like if we wanted to use multiple transforms. But we could still check data with. multiple partition specs in it (possibly doing 3 columns worth of changes).

Fully up to you.

Reply:
I made sure this particular test (just made some mods) will necessarily make 320+ files if the shuffle isn't appropriately applied. I'm going to add some more tests in a follow issue to deal with the partition grouping and V2 tables.